encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/TypeTokenContainer.java

```diff
@@ -15,13 +15,33 @@
 package com.google.firebase.decoders;
 
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import java.lang.reflect.Type;
 
 /**
  * {@link TypeTokenContainer} is used to get actual type parameter in a generic class at given
  * index.
  */
-// TODO: change from interface to class
-public interface TypeTokenContainer {
+public class TypeTokenContainer {
+  private final TypeToken<?>[] typeTokens;
+
+  TypeTokenContainer(@Nullable Type[] types) {
```

Comment:
Consider moving the conversion logic `Type->TypeToken` into the TypeToken class itself, this would increase cohesion of this class and will not require making `TypeToken#of(Type)` non-private, effectively exposing its implementation details.

Reply:
Also might make sense to make the constructor public and class `final`.